### PR TITLE
Spark: Support metadata columns in 3.2

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds proper support for metadata columns in Spark 3.2.

### CODE REVIEW

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java
Comment (wypoon) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java:184]: I am not too familiar with the metadata columns, but I see 5 currently defined in Iceberg's `MetadataColumns`. Is there a reason to omit `_deleted` here? And it probably doesn't matter, but should we keep the columns in the order defined in `MetadataColumns` (ordered by id from -1 to -5)?

**Review by wypoon — COMMENTED** [review]
LGTM otherwise.

One other question: In `SupportsMetadataColumns`, it says "If a table column and a metadata column have the same name, the metadata column will never be requested. It is recommended that Table implementations reject data column name that conflict with metadata column names." Do we have any logic that does that (reject data column names that conflict with metadata column names)? If not, we should, right?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java
Comment (aokolnychyi) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java:184]: I can match the order in `MetadataColumns` for consistency.

I am not sure how useful `_deleted` metadata column will be in Spark now. I guess it will be always `false`?
@jackye1995 @chenjunjiedada @RussellSpitzer @rdblue, any thoughts?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java
Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java:182]: I like that we can project the partition. I've been meaning to add a way to project the individual partition fields, but this is probably way easier.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java
Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java:184]: `_deleted` can be added later. The purpose of that field is to allow us to merge deletes in actions.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks great. Thanks, @aokolnychyi!

**Comment by rdblue** [comment]
> If a table column and a metadata column have the same name, the metadata column will never be requested. Currently we have very simple names for Iceberg metadata columns. I wonder if we should make it more complex on engine side, such as _iceberg_partition instead of just _partition. Any thoughts?

Sorry I didn't see this ongoing discussion before I merged. If everyone is okay with it, let's continue to discuss and address it in a follow-up. At least that way we unblock Anton's other work.

I would be fine adding a way to detect this case and change the column name. Or just waiting for someone to complain. I doubt many people are using `_partition` so it may be better just to wait until someone hits it in practice.

**Pull request merged**
